### PR TITLE
95690 GetOutstandingIPO optimize query and add index

### PR DESCRIPTION
**Pull request by krsteen**
Optimized query using projection.
Added two new indexes on Invitations and Participant to support new queries

[AB#95690](https://statoildeveloper.visualstudio.com/b6d97460-2e53-47ae-aad4-b41396d90828/_workitems/edit/95690)

### CODE REVIEW

File: src/Equinor.ProCoSys.IPO.Infrastructure/IPOContext.cs [hunk 1]
```diff
@@ -37,6 +38,30 @@ protected override void OnModelCreating(ModelBuilder modelBuilder)
             base.OnModelCreating(modelBuilder);
             modelBuilder.ApplyConfigurationsFromAssembly(Assembly.GetExecutingAssembly());
             SetGlobalPlantFilter(modelBuilder);
+            SetIndexes(modelBuilder);
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Infrastructure/IPOContext.cs:41]: Undo changes in this file and move index creation to existing ParticipantConfiguration and existing InvitationConfiguration

File: src/Equinor.ProCoSys.IPO.Query/GetOutstandingIpos/GetOutstandingIposForCurrentPersonQueryHandler.cs [hunk 2]
```diff
@@ -31,34 +32,86 @@ public class GetOutstandingIposForCurrentPersonQueryHandler : IRequestHandler<Ge
             _plantProvider = plantProvider;
         }
 
+        public async Task<List<InvitationDto>> GetNonCanceledInvitations(CancellationToken cancellationToken)
+        {
+
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Query/GetOutstandingIpos/GetOutstandingIposForCurrentPersonQueryHandler.cs:37]: remove blank

File: src/Equinor.ProCoSys.IPO.Query/GetOutstandingIpos/GetOutstandingIposForCurrentPersonQueryHandler.cs [hunk 3]
```diff
@@ -31,34 +32,86 @@ public class GetOutstandingIposForCurrentPersonQueryHandler : IRequestHandler<Ge
             _plantProvider = plantProvider;
         }
 
+        public async Task<List<InvitationDto>> GetNonCanceledInvitations(CancellationToken cancellationToken)
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Query/GetOutstandingIpos/GetOutstandingIposForCurrentPersonQueryHandler.cs:35]: private and move method below Handle()

File: src/Equinor.ProCoSys.IPO.Query/GetOutstandingIpos/GetOutstandingIposForCurrentPersonQueryHandler.cs [hunk 4]
```diff
@@ -31,34 +32,86 @@ public class GetOutstandingIposForCurrentPersonQueryHandler : IRequestHandler<Ge
...
+                             Organization = p.Organization,
+                             SignedAtUtc = p.SignedAtUtc,
+                             SortKey = p.SortKey,
+                             Type = p.Type,
+                             SignedBy = p.SignedBy
+
+                         })
+                         .ToListAsync(cancellationToken);
+
+            var nonCancelledInvitationsGrouped = new List<InvitationDto>();
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Query/GetOutstandingIpos/GetOutstandingIposForCurrentPersonQueryHandler.cs:60]: rename and rearange to this to ease reading
image

File: src/Equinor.ProCoSys.IPO.Query/GetOutstandingIpos/GetOutstandingIposForCurrentPersonQueryHandler.cs [hunk 5]
```diff
@@ -31,34 +32,86 @@ public class GetOutstandingIposForCurrentPersonQueryHandler : IRequestHandler<Ge
...
 
-                var currentUsersOutstandingInvitations = new List<Invitation>();
                 var listHasFunctionalRoles =
-                    nonCancelledInvitations.Any(i => i.Participants.Any(p => p.FunctionalRoleCode != null));
-                IList<string> currentUsersFunctionalRoleCodes = new List<string>();
-                if (listHasFunctionalRoles)
-                {
-                    currentUsersFunctionalRoleCodes = await _meApiService.GetFunctionalRoleCodesAsync(_plantProvider.Plant);
-                }
+               nonCancelledInvitationsGrouped.Any(i => i.Participants.Any(p => p.FunctionalRoleCode != null));
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Query/GetOutstandingIpos/GetOutstandingIposForCurrentPersonQueryHandler.cs:109]: indent

File: src/Equinor.ProCoSys.IPO.Query/GetOutstandingIpos/GetOutstandingIposForCurrentPersonQueryHandler.cs [hunk 6]
```diff
@@ -93,30 +146,30 @@ public class GetOutstandingIposForCurrentPersonQueryHandler : IRequestHandler<Ge
...
                 && p.Organization != Organization.External
-                && p.SortKey != 1) 
+                && p.SortKey != 1)
                 || (p.SortKey == 1 && invitation.Status == IpoStatus.Completed)));
 
-        private static bool UserWasInvitedAsPersonInFunctionalRole(Invitation invitation, IEnumerable<string> currentUsersFunctionalRoleCodes)
+        private static bool UserWasInvitedAsPersonInFunctionalRole(InvitationDto invitation, IEnumerable<string> currentUsersFunctionalRoleCodes)
         {
-            var functionalRoleParticipantCodesOnInvitation = 
+            var functionalRoleParticipantCodesOnInvitation =
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Query/GetOutstandingIpos/GetOutstandingIposForCurrentPersonQueryHandler.cs:161]: rearrange the last part to ease reading
image

File: src/Equinor.ProCoSys.IPO.Query/GetOutstandingIpos/GetOutstandingIposForCurrentPersonQueryHandler.cs [hunk 7]
```diff
@@ -70,19 +123,19 @@ public class GetOutstandingIposForCurrentPersonQueryHandler : IRequestHandler<Ge
...
+                               p.SignedBy == null &&
+                               (p.AzureOid == currentUserOid ||
+                               currentUsersFunctionalRoleCodes.Contains(p.FunctionalRoleCode))).Organization;
+                           return new OutstandingIpoDetailsDto
+                           {
+                               InvitationId = invitation.Id,
+                               Description = invitation.Description,
+                               Organization = organization
+                           };
+                       }));
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Query/GetOutstandingIpos/GetOutstandingIposForCurrentPersonQueryHandler.cs:138]: Below there is a catch Exception which swallow the exception silently
1) Log the exception
2) Create a TECH-PBI, where we alter the logic to still log the exception **and** return 500-error. The client (Main) must handle this error and show a error-message like "something went wrong when getting outstanding ipos". 
Write in TECH-PBI: Remember to check if any else ipo endpoint used by Main catch Exceptions which swallow the exception silently

**Review by janid1967 — CHANGES_REQUESTED** [review]
See already added comments

File: src/Equinor.ProCoSys.IPO.Query/GetOutstandingIpos/GetOutstandingIposForCurrentPersonQueryHandler.cs (hunk 7 above)
Comment (janid1967) [reply src/Equinor.ProCoSys.IPO.Query/GetOutstandingIpos/GetOutstandingIposForCurrentPersonQueryHandler.cs:138]: Agree @bratli ?

Comment (bratli) [reply src/Equinor.ProCoSys.IPO.Query/GetOutstandingIpos/GetOutstandingIposForCurrentPersonQueryHandler.cs:138]: Yes

Comment (krsteen) [reply src/Equinor.ProCoSys.IPO.Query/GetOutstandingIpos/GetOutstandingIposForCurrentPersonQueryHandler.cs:138]: TECH PBI Created; 96013

**Review by janid1967 — APPROVED** [review]

**Review by janid1967 — CHANGES_REQUESTED** [review]
Failed tests

**Review by janid1967 — APPROVED** [review]

**Pull request merged**